### ISSUE TITLE
Nested resource paths; possible?

### ISSUE THREAD

**Issue by sverhagen** [eclipse/californium#1371]
Using version 2.2 of Californium. Thank you for this library.

I'm trying to make nested resource paths, hoping for something like so: `coaps://localhost/api/foo/bar`. I think I understand I could make an `api` resource with a `foo` child with a `bar` child. But I find it a bit convoluted, and am wondering if there's a way to just express this in the name, like:

```java
public class BarResource extends CoapResource {
    public BarResource() {
        super("api/foo/bar" /* name */);
    }
}
```

This does obviously not work. I have also tried using the `setPath` method, which I realize may not be meant for that; it's getting reset in `setParent` that's called later. It seems like `getURI` may be more what I want, but it's value is only derived of the others.

In other words, this is probably not how this is supposed to work, if at all? Thanks for any pointers.

In a more generalized form of asking the question, is there any conceptual documentation on `name`, `path` and `uri` properties of a resource?

**Comment by boaks** [comment]
> In a more generalized form of asking the question, is there any conceptual documentation on name, path and uri properties of a resource?

I'm not sure, what that document should contain. The API is from 2015, the initial implementation phase at the ETH. For "read access" it seem to be quiet common, very similar to the java File API:

- name the name of the node itself
- path the list of names of parents
- uri path + uri

The "write access" may be surprising, for me the "setPath" looks pretty much for a "internal function", used by the "setParent". Generally, Californium doesn't process the resources as map of URI (or URI patterns), it uses a resource tree. 

> But I find it a bit convoluted, and am wondering if there's a way to just express this in the name, like:

What would you consider then as result of GET /api/foo ? NOT FOUND? Or a 2.05 without payload? I'm not sure, if the answer would be the same for every use-case. That would make a solution with a give use-case simple, but a general would require more work.

**Comment by sverhagen** [comment]
Thank you for the quick reply.

>path the list of names of parents

That makes sense, except that I can't seem to set it, which you address:

>The "write access" may be surprising, for me the "setPath" looks pretty much for a "internal function"

That is fair. But a little confusing (maybe that's what you meant with "surprising") that it is `public` and also on the `Resource` interface.

>Californium doesn't process the resources as map of URI (or URI patterns), it uses a resource tree

I guess that is really my answer. After inspecting the code I guessed that this was the case, I just wanted to make sure.

>What would you consider then as result of GET /api/foo ? NOT FOUND?

Yes, that was my expectation. Just for some context, the bulk of my work is in Spring, using their `RequestMapping` and friends, which would treat `/api/foo` as much of a "HTTP 404 Not Found" as `/api/something`, given there is an `@RequestMapping("/api/foo/bar")`. I think either approach has its own historical origin, and that is fine. Still for context, this question just came from my broader effort to make our _one_ Californium-based service similar to our many Spring services.

**Comment by boaks** [comment]
> 404

Let me guess, that DELETE "/api/foo" should also not delete "something". Others may required to delete all child nodes. I'm sure, you understand, that for a give use-case it may be easy to provide a implementation (I guess, you will just create a helper for that), but for general usage, it would get laaarge :-).

Any reasons to use 2.2.0? There is a 2.2.3 and a 2.3.0.

**Comment by sverhagen** [comment]
>Any reasons to use 2.2.0? There is a 2.2.3 and a 2.3.0.
No reason, just not having a very well-defined practice of regularly updating the dependencies.

Thanks for your help!
